GitUI/CommandsDialogs/CustomDiffMergeTool.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel;
+using System.Drawing;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using GitCommands;
+
+namespace GitUI
+{
+    public class CustomDiffMergeTool
```

Comment:
I have multiple issues with the API naming and coherency in this type.

E.g. `CustomDiffMergeTool` implies a singular item. `Clear()` generally implies a collection.
I also struggle with `LoadCustomDiffMergeTools(...)` method in this class, it feels misplaced.

Reply:
`CustomDiffMergeTool` is a data class
`LoadCustomDiffMergeTools` is a shared method for the GUI. You asked for a common implementation so I put it as a static method there. The cache was added in the latest change (I did not want to add such functionality here and adding to GitModule or the cache would be simpler but we do not want to add functionality there.

Reply:
I really struggle with this API. It mixes multiple concerns - it is a data class, and it also responsible for data load, data parsing, UI binding, and cache busting.
It also appears completely divorced from diff/merge tool API that live in GitCommands.DiffMergeTools namespace.

I spent some time (I'm quite time poor atm) thinking how I would approach the plumbing, and come up with this draft: https://github.com/RussKie/gitextensions/commit/caadf00366bd31330330c679e62ac2f6390101f0

In a nutshell, I propose moving data load and parsing into a dedicated class in GitCommands.DiffMergeTools namespace: `CustomDiffMergeToolProvider`. It can also provide a caching mechanism, as caching isn't something its callers should be concerned with. I propose we do a simple caching based in a working directory, as I can't think of a scenario where we have multiple concurrent calls from different repositories in the same process (those calls ultimately will be coming from a UI). NB: I can be totally overlooking some case without realising it.
The UI layer is only responsible for binding the list of custom items, nothing else.

Reply:
> I really struggle with this API. It mixes multiple concerns - it is a data class, and it also responsible for data load, data parsing, UI binding, and cache busting.
> It also appears completely divorced from diff/merge tool API that live in GitCommands.DiffMergeTools namespace.
> 
> I spent some time (I'm quite time poor atm) thinking how I would approach the plumbing, and come up with this draft: [RussKie@caadf00](https://github.com/RussKie/gitextensions/commit/caadf00366bd31330330c679e62ac2f6390101f0)
> 
> In a nutshell, I propose moving data load and parsing into a dedicated class in GitCommands.DiffMergeTools namespace: `CustomDiffMergeToolProvider`. It can also provide a caching mechanism, as caching isn't something its callers should be concerned with. I propose we do a simple caching based in a working directory, as I can't think of a scenario where we have multiple concurrent calls from different repositories in the same process (those calls ultimately will be coming from a UI). NB: I can be totally overlooking some case without realising it.
> The UI layer is only responsible for binding the list of custom items, nothing else.

This _is_ separate from GE DiffMergeTools that offer tools to Git. The custom tolls lists the tools Git provides.

I would prefer to not use `CustomDiffMergeToolProvider`, the menu setup in LoadCustomDiffMergeTools() is handling `IContainer components` to make sure the ContextMenuStrip is disabled. This is why it is in GitUI. But I can change that if you prefer.

For caching, it takes some time to run the Git command (as discussed with mstv above). This need to be changed in master also if this PR is not merged (by moving CustomDiffMergeTool implementation to RevDiff and Resolve conflicts). As this is such a slow command, I do not want to add this to every time a new repo is opened. The possibility to add a difftool for a repo is something I would like to ignore.
However, as a user I would expect that adding a difftool in GE would take effect immediately.

Locking is needed as several forms wants access the same information. Of course that can be documented too.
(I would like to implement this in the GitCommandCache instead, to have a semaphore on a command with cache requested already running, but there are some caveeats there.)